scarlet/source.py

```diff
@@ -341,108 +310,59 @@ def __init__(self, frame, sky_coord, observation, symmetric=True, monotonic=True
+            if np.all(sed <= 0):
+                logger.warning(msg)
             else:
-                self._centroid_weight = self.frame.psfs[0].image
-
-        # ensure adherence to constraints
-        self.update()
-
-    def update(self):
-        """Default update parameters for an ExtendedSource
-
-        This method can be overwritten if a different set of constraints
-        or update functions is desired.
-        """
-        if self._parent is None:
-            it = 0
-        else:
-            it = self._parent.it
-
-        # Update the central pixel location (pixel_center)
-        self.pixel_center = measurement.max_pixel(self.morph, self.pixel_center)
-
-        # Thresholding needs to be fixed (DM-10190)
-        # if it > self.delay_thresh:
-        #     update.threshold(self)
+                logger.info(msg)
 
-        # If there is a threshold bounding box, use it
-        if hasattr(self, "bboxes") and "thresh" in self.bboxes:
-            bbox = self.bboxes["thresh"]
-        else:
-            bbox = None
+        # set up parameters
+        sed = Parameter(sed, step=partial(relative_step, factor=1e-2), constraint=PositivityConstraint())
+        center = Parameter(self.center, step=1e-1)
 
-        if self.symmetric:
-            # Update the centroid position
-            if it % 5 == 0:
-                self.pixel_center, self.shift = measurement.psf_weighted_centroid(self.morph, self._centroid_weight, self.pixel_center)
+        # define bbox
+        pixel_center = tuple(np.round(center).astype('int'))
+        front, back = 0, C
+        bottom = pixel_center[0] - frame.psf.shape[1]//2
+        top = pixel_center[0] + frame.psf.shape[1]//2
+        left = pixel_center[1] - frame.psf.shape[2]//2
+        right = pixel_center[1] + frame.psf.shape[2]//2
+        bbox = Box.from_bounds(front, back, bottom, top, left, right)
 
-            # make the morphology perfectly symmetric
-            update.symmetric(self, self.pixel_center, algorithm="kspace", bbox=bbox)
+        super().__init__(frame, sed, center, self._psf_wrapper, bbox=bbox)
 
-        if self.monotonic:
-            # make the morphology monotonically decreasing
-            update.monotonic(self, self.pixel_center, bbox=bbox)
+    def _psf_wrapper(self, *parameters):
+        return self.frame.psf.__call__(*parameters, bbox=self.bbox)[0]
 
-        update.positive(self)  # Make the SED and morph non-negative
-        update.normalized(self)  # Use MORPH_MAX normalization
-        return self
 
 
-class ExtendedSource(PointSource):
-    def __init__(self, frame, sky_coord, observation, bg_rms, thresh=1,
-                 symmetric=True, monotonic=True, center_step=5, delay_thresh=10, **component_kwargs):
+class ExtendedSource(FactorizedComponent):
+    def __init__(self, frame, sky_coord, observations, obs_idx=0, thresh=1.,
+                 symmetric=True, monotonic=True, shifting=False):
```

Comment:
please make `symmetric=False` the default.

Reply:
Also I would optionally allow the user to specify a `ConstraintChain` instead of only allowing certain constraints. It would make the model more flexible.

Reply:
See arguments above: the type of constraint is essential to fully specify the model. It should not live in user space. It's not reproducible and probably can't be pickled.